### PR TITLE
feat: incoming request validation

### LINKED ISSUE #6: Validate incoming DataVaultConfigurations and Documents
Currently the only check is a basic JSON unmarshal call. This means that the only check the server will do right now is validating whether the input is valid JSON, but not whether it's actually got the correct data per the spec. Implement robust validation.

~~Additional check: return an error if an "invalid id" is provided. The ID has to work correctly when concatenated to a URL. Any provided ID that will cause this to fail (such as ones that contain `//`) need to generate an error.~~ No longer applicable since vault IDs are generated by the server, and document IDs have to be Base58-encoded.

### PR DESCRIPTION
**Pull request by grace0925**
Closes #6 
Signed-off-by: Grace Liu <grace.liu@securekey.com>

### CODE REVIEW

File: pkg/restapi/messages/messages.go [hunk 1]
```diff
@@ -48,9 +48,35 @@ Received data: %s`
...
+	InvalidControllerString = "invalid controller value: %w"
+	// InvalidInvokerStringArray is the message returned by the EDV server when a attempt is made to create a vault
+	// with invalid invoker values.
+	InvalidInvokerStringArray = "invalid invoker value: %w"
+	// InvalidDelegatorStringArray is the message returned by the EDV server when a attempt is made to create a vault
+	// with invalid delegator values.
+	InvalidDelegatorStringArray = "invalid delegator value: %w"
+	// InvalidKEKIDString is the message returned by the EDV server when a attempt is made to create a vault
+	// with an invalid key agreement key id value.
+	InvalidKEKIDString = "invalid key agreement key id: %w"
```
Comment (DRK3) [inline pkg/restapi/messages/messages.go:79]: `id` should be capitalized in the messages

File: pkg/restapi/operation/operations.go [hunk 2]
```diff
@@ -457,3 +460,80 @@ func (vc *VaultCollection) queryVault(vaultID string, query *models.Query) ([]st
...
+		return fmt.Errorf(messages.InvalidControllerString, err)
+	}
+
+	if err := checkFieldsWithURIArray(&dataVaultConfig.Invoker, dataVaultConfig.Controller); err != nil {
+		return fmt.Errorf(messages.InvalidInvokerStringArray, err)
+	}
+
+	if err := checkFieldsWithURIArray(&dataVaultConfig.Delegator, dataVaultConfig.Controller); err != nil {
+		return fmt.Errorf(messages.InvalidDelegatorStringArray, err)
+	}
```
Comment (DRK3) [inline pkg/restapi/operation/operations.go:479]: I don't think we should be modifying the actual data vault configuration being passed in. Rather, if/when we add code later that uses those Controller, Invoker and Delegator fields, we can have the logic there that determines which one to use based on which ones are empty. This way, the original data vault configuration is preserved. This way, we can differentiate between a data vault configuration that was purposefully initialized with all three fields being the same value (which would be redundant) and one where we just default to the controller field. While your approach may technically work the same way, I think it's safer to leave the config as is and defer the logic of what to do if one thing is empty until we actually need to check.

File: pkg/restapi/operation/operations.go [hunk 3]
```diff
@@ -457,3 +460,80 @@ func (vc *VaultCollection) queryVault(vaultID string, query *models.Query) ([]st
...
+	if encodingErr := edvutils.CheckIfBase58Encoded128BitValue(doc.ID); encodingErr != nil {
+		return encodingErr
+	}
+
+	if doc.JWE == nil {
+		return errors.New(messages.BlankJWE)
+	}
+
+	return nil
+}
```
Comment (DRK3) [inline pkg/restapi/operation/operations.go:539]: I think this logic applies to _any_ encrypted document, not just incoming ones, so you can remove the `Incoming` part of the function name.

File: pkg/restapi/operation/operations.go [hunk 4]
```diff
@@ -457,3 +460,80 @@ func (vc *VaultCollection) queryVault(vaultID string, query *models.Query) ([]st
...
+
+func validateIncomingEncryptedDocument(doc models.EncryptedDocument) error {
+	if encodingErr := edvutils.CheckIfBase58Encoded128BitValue(doc.ID); encodingErr != nil {
+		return encodingErr
+	}
+
+	if doc.JWE == nil {
+		return errors.New(messages.BlankJWE)
+	}
```
Comment (DRK3) [inline pkg/restapi/operation/operations.go:536]: In addition to checking that it's not nil, you should also make sure it's not just an empty `json.RawMessage`.

I think there's something else we should check in the JWE at a minimum... @Baha-sk, @troyronda, what else in the JWE should we check? I _think_ someone told me once that we should check the "alg" field?

File: pkg/restapi/operation/operations_test.go [hunk 5]
```diff
@@ -684,7 +706,33 @@ func TestCreateDocument(t *testing.T) {
 		createDocumentEndpointHandler.Handle().ServeHTTP(rr, req)
 
 		require.Equal(t, http.StatusBadRequest, rr.Code)
-		require.Equal(t, fmt.Sprintf(messages.CreateDocumentFailure, vaultID, messages.ErrNot128BitValue),
+		require.Equal(t, fmt.Sprintf(messages.InvalidDocument, vaultID, messages.ErrNot128BitValue),
+			rr.Body.String())
+	})
+	t.Run("Empty jwe", func(t *testing.T) {
```
Comment (DRK3) [inline pkg/restapi/operation/operations_test.go:712]: `jwe` should be capitalized

**Review by DRK3 — CHANGES_REQUESTED** [review]

File: pkg/restapi/operation/operations.go (hunk 2 above)
Comment (llorllale) [reply pkg/restapi/operation/operations.go:479]: @DRK3 

> I  don't think we should be modifying the actual data vault configuration being passed in

I think you meant "verifying" right?

> if/when we add code later that uses those Controller, Invoker and Delegator fields

:+1:

File: pkg/restapi/operation/operations.go (hunk 4 above)
Comment (llorllale) [reply pkg/restapi/operation/operations.go:536]: @DRK3 

> In addition to checking that it's not nil, you should also make sure it's not just an empty json.RawMessage.

`json.RawMessage` is a typed `[]byte`. It's enough to just check for length.

> I think there's something else we should check in the JWE at a minimum... @Baha-sk, @troyronda, what else in the JWE should we check? I think someone told me once that we should check the "alg" field?

Must ban `none`.

File: pkg/edvutils/edvutils_test.go [hunk 6]
```diff
@@ -20,6 +21,9 @@ const (
 	testConvertedUUIDString       = "d15034fa-9525-4ebf-3352-d19c8b02cf05"
 
 	not128BitString = "testString"
+
+	validURIString   = "did:example:123456789"
+	invalidURIString = "invalidURI"
```
[the remark this answers is not in the document]
Comment (llorllale) [reply pkg/edvutils/edvutils_test.go:26]: +1 to move away from hungarian notation

File: pkg/restapi/operation/operations.go (hunk 2 above)
Comment (DRK3) [reply pkg/restapi/operation/operations.go:479]: @llorllale I did mean "modifying" - that method actually modifies the Data Vault Configuration.

Comment (llorllale) [reply pkg/restapi/operation/operations.go:479]: Oh. That's _very_ bad practice and unexpected. In general, refrain from mutating input args.

**Pull request closed without merging**